src/bin/pg_autoctl/supervisor.c

```diff
@@ -529,6 +469,77 @@ supervisor_stop(Supervisor *supervisor)
 }
 
 
+/*
+ * If we have received a signal that instructs a shutdown, such as SIGTERM or
+ * SIGINT, then we need to do one of these things:
+ *
+ * - first time we receive the signal, begin a shutdown sequence for all
+ *   services and the main supervisor itself,
+ *
+ * - when receiving the signal again, if it's a SIGTERM continue the shutdown
+ *   sequence,
+ *
+ * - when receiving a SIGINT forward it to our services so as to finish as fast
+ *   as we can, and from then on always use SIGINT (to that end we use
+ *   supervisor->shutdownSignal)
+ *
+ * Sending SIGTERM and then later SIGINT if the process is still running is a
+ * classic way to handle service shutdown.
+ */
+static void
+supervisor_handle_signals(Supervisor *supervisor)
+{
+	bool shutdownNow = false;
+	int signal = asked_to_stop_fast ? SIGINT : SIGTERM;
+	const char *signalStr = asked_to_stop_fast ? "SIGINT" : "SIGTERM";
+
+	if (asked_to_stop || asked_to_stop_fast)
+	{
+		/* use new signal, unless we were already using SIGINT */
+		if (supervisor->shutdownSignal != SIGINT)
+		{
+			supervisor->shutdownSignal = signal;
+		}
+
+		if (!supervisor->shutdownSequenceInProgress)
+		{
+			log_info("pg_autoctl received signal %s, terminating", signalStr);
+
+			supervisor->cleanExit = true;
+			supervisor->shutdownSequenceInProgress = true;
+
+			shutdownNow = true;
+		}
+		else if (asked_to_stop_fast)
+		{
+			/*
+			 * Even when we are already shutting down, we process SIGINT again
+			 * and log about it.
+			 */
+			log_info("pg_autoctl received signal SIGINT, terminating");
+
+			shutdownNow = true;
+		}
+	}
+
+	/*
+	 * Initiate the stop sequence the first time we receive SIGTERM and then
+	 * each time we receive SIGINT.
+	 */
+	if (shutdownNow)
+	{
+		/* stop all the services. */
+		(void) supervisor_stop_subprocesses(supervisor);
+
+		/* allow for processing SIGINT again next time it's sent: reset */
+		if (asked_to_stop_fast)
```

Comment:
I think there might still be a race condition if you send both SIGTERM and SIGINT very closely together. I think this should fix it, since we check which signal we handled this time, not which flags are set. 

```suggestion
		if (signal == SIGINT)
```

Otherwise `asked_to_stop_fast` might be set to true after setting `signal`. This way we do not unset the `asked_to_stop_flag` unless we actually handled it right now. If we did not handle it, it will simply be handled next loop, because it's still set.